### PR TITLE
Apply `type_union_resolution` to array and values

### PR DESCRIPTION
**Pull request by jayzhan211**
## Which issue does this PR close?

Closes #.

## Rationale for this change

* Main: Follow on https://github.com/apache/datafusion/pull/10268. Apply `type_union_resolution ` to Array and Values coercion

* Cleanup: Remove `coerce_arguments_for_fun` and move the coercion logic to `coerce_types`. So does the null to i64 conversion.

### CODE REVIEW

File: datafusion/sqllogictest/test_files/map.slt [hunk 1]
```diff
@@ -148,18 +148,17 @@ SELECT MAKE_MAP([1,2], ['a', 'b'], [3,4], ['b']);
 {[1, 2]: [a, b], [3, 4]: [b]}
 
 query ?
-SELECT MAKE_MAP('POST', 41, 'HEAD', 'ab', 'PATCH', 30);
+SELECT MAKE_MAP('POST', 41, 'HEAD', 53, 'PATCH', 30);
```
Comment (jayzhan211) [inline datafusion/sqllogictest/test_files/map.slt:151]: It seems mixed type is not allowed in duckdb too

File: datafusion/expr/src/logical_plan/builder.rs [hunk 2]
```diff
@@ -209,7 +209,9 @@ impl LogicalPlanBuilder {
                 }
                 if let Some(prev_type) = common_type {
                     // get common type of each column values.
-                    let Some(new_type) = values_coercion(&data_type, &prev_type) else {
+                    let data_tyeps = vec![prev_type.clone(), data_type.clone()];
+                    let Some(new_type) = type_union_resolution(&data_tyeps) else {
+                        println!("data_tyeps: {:?}", data_tyeps);
```
Comment (seckin) [inline datafusion/expr/src/logical_plan/builder.rs:214]: do we want this here?

File: datafusion/sqllogictest/test_files/select.slt [hunk 3]
```diff
@@ -348,17 +348,23 @@ VALUES (1),()
 statement error DataFusion error: Error during planning: Inconsistent data length across values list: got 2 values in row 1 but expected 1
 VALUES (1),(1,2)
 
-statement error DataFusion error: Error during planning: Inconsistent data type across values list at row 1 column 0
+query I
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/select.slt:351]: 👍 

Looks like an improvement to me

**Review by alamb — APPROVED** [review]
Thank you @jayzhan211  and @seckin  -- this looks like an improvement to me

I think @seckin 's comment about the extra `println` should be addressed prior to merging this PR. The typo would be great to fix too

**Comment by jayzhan211** [comment]
Thanks @alamb @seckin

**Pull request merged**